Editorial review: Document Wasm bulk memory operations (#44296)
mdn/content: files/en-us/webassembly/reference/memory/copy/index.md

- `index`
  - : The `memory`'s index number, for example `0` for the first `memory` in the wasm module, `1` for the second, etc.

If `dest_memory` or `source_memory` are omitted, they default to `0`.

### Type

```plain
[dest_offset source_offset length] -> []
```

- `dest_offset`
  - : An integer representing the offset to start writing the copied data to, in the destination memory. This will be an [`i32`](/en-US/docs/WebAssembly/Reference/Value_types/i32) or an [`i64`](/en-US/docs/WebAssembly/Reference/Value_types/i64), to match the [`address_type`](/en-US/docs/WebAssembly/Reference/Definitions/memory#address_type) the `memory` was defined with.
- `source_offset`
  - : An integer representing the offset to start copying data from, in the source memory. This will be an `i32` or an `i64`, to match the `address_type` the `memory` was defined with.
- `length`
  - : An integer representing the number of bytes to copy. This will be an `i32` or an `i64`, to match the `address_type` the `memory` was defined with. When copying between a 32-bit and 64-bit memory, an `i32` must be used for the `length`.

> [!NOTE]

> > `memory.copy` copies memory regions in an overlap-aware way. In other words, the source data identified by `source_offset` and `length` is first copied into a temporary value before being copied starting at `dest_offset`, meaning that if your source and destination data overlap, they don't interfere with one another and the source is cleanly copied into the destination region as expected.

Reviewer: Pedantic suggestion:

```suggestion
> `memory.copy` copies memory regions in an overlap-aware way. In other words, the copy proceeds as if the source data identified by `source_offset` and `length` was first copied into a temporary value before being copied to `dest_offset`, meaning that if your source and destination data overlap, they don't interfere with one another and the source is cleanly copied into the destination region as expected.
```

In reality there need not be a temporary value; typical implementations will just check for overlap and then decide whether to copy bytes front-to-back or back-to-front.

Reviewer: Could be worth noting that `table.copy` works exactly the same way. But, nobody ever uses table.copy anyhow, so who cares.

Author: I've used your suggestion for the `memory.copy` note. I also copied it to `table.copy` and updated the wording to make sense for tables. You never know, someone might care ;-)

### Traps

If any copied byte would be out of bounds in the source or destination, the instruction traps.

### Binary encoding

| Instruction   | Binary format                   | Example text => binary                     |
| ------------- | ------------------------------- | ------------------------------------------ |
| `memory.copy` | `0xfc 10:u32 x:memidx x:memidx` | `memory.copy 0 0` => `0xfc 0x0a 0x00 0x00` |

## Examples

> ### Copy within default memory

Reviewer 2: I didn't immediately realize by default memory you meant "the zero index memory" vs "copy within a specified memory" so this, and https://github.com/mdn/content/pull/44296/changes#r3654129806 and same problem for fill and grow docs

```suggestion
### Copy within the default (0-index) memory
```

Author: Makes sense. I've fixed all of them up.

sidebar: webassemblysidebar
---

> The **`memory.copy`** [memory instruction](/en-US/docs/WebAssembly/Reference/Memory) copies data from one [`memory`](/en-US/docs/WebAssembly/Reference/Definitions/memory) region to another.

Reviewer: Possibly worth saying that this instruction is analogous to [`memmove`](https://en.cppreference.com/cpp/string/byte/memmove) in C, in other words that it copies a region of memory in an overlap-aware way. 

For example, given the data "Hello Wasm", doing `(memory.copy (i32.const 2) (i32.const 0) (i32.const 5))` will result in "HeHelloasm", whereas a naive byte-by-byte copy might inadvertently produce "HeHeHeHasm", because the source overlapped with the destination.

Author: OK, I've added a line to explain the analogy, in the same way as with `memory.fill` and `memset`.

I've also added a note further down the page to explain the overlap-aware copying. Let me know if you think it reads OK.

{{InteractiveExample("Wat Demo: memory.copy", "tabbed-taller")}}

```wat interactive-example
(module
  (memory $memory1 1)

>   (memory $memory2 (export "memory") 1)

Reviewer: This is a good example, but multi-memory is very niche. Almost 100% of the time you only have one memory. So I wonder if probably the "copy within default memory" example from below should be brought up here and adapted (e.g. to print "Hello Hello").

Author: This makes sense. I've updated the live example to use one memory, and updated the example description to suit.

  (data $greeting (memory $memory1) (i32.const 0) "Hello World")

  (func (export "copy")
    i32.const 0       ;; destination memory offset
    i32.const 6       ;; source memory offset

>     i32.const 5      ;; number of bytes to copy

Reviewer: ```suggestion
    i32.const 5       ;; number of bytes to copy
```

Author: "shoddy", as we would say in the UK. Fixed now ;-)

Reviewer 2: It's so hard to find good help now days :-)

    memory.copy $memory2 $memory1
  )
)
```

```js interactive-example
WebAssembly.instantiateStreaming(fetch("{%wasm-url%}")).then((result) => {
  result.instance.exports.copy();
  const memBuffer = result.instance.exports.memory.buffer;
  const memArray = new Uint8Array(memBuffer, 0, 5);
  console.log(new TextDecoder().decode(memArray));
});
```

In the above example, we define two memories: `$memory1` and `$memory2`, both with a maximum size of one page. The second memory is exported. We then include:

- A [`data`](/en-US/docs/WebAssembly/Reference/Definitions/data) definition that contains the string "Hello World" and immediately copies the entire string into `$memory1`.
- An exported `copy()` function that copies the string "World" from `$memory1` into `$memory2`.

In the JavaScript, we call the `copy()` function to populate the exported memory with data, then decode the exported memory's buffer and log the result to the console.

## Syntax

```plain
memory.copy dest_memory source_memory
```

- `memory.copy`
  - : The `memory.copy` instruction type. Must always be included first.

> - `dest_memory` {{optional_inline}}
>   - : The identifier for the `memory` you want to copy the data into.
> - `source_memory` {{optional_inline}}
>   - : The identifier for the `memory` you want to copy the data from.

Reviewer: Nit: I generally suggest describing things in terms of indexes, because identifiers are just sugar for indexes.

```suggestion
- `dest_memory` {{optional_inline}}
  - : The index of the `memory` you want to copy the data into.
- `source_memory` {{optional_inline}}
  - : The index of the `memory` you want to copy the data from.
```

Author: I'm not sure I agree with this, as it clashes with my naming conventions: the identifier being a name or an index.

If you feel strongly about this, could we instead add something to the `name` definition to state that it is syntactical sugar for the underlying index number?

`dest_memory` and `source_memory` can be one of the following:

- `name`
  - : An identifying name [set for the `memory`](/en-US/docs/WebAssembly/Reference/Definitions/memory#name) when it was first defined. This must begin with a `$` symbol, for example `$my_mem`.
- `index`
  - : The `memory`'s index number, for example `0` for the first `memory` in the wasm module, `1` for the second, etc.

If `dest_memory` or `source_memory` are omitted, they default to `0`.

### Type

```plain
[dest_offset source_offset length] -> []
```

- `dest_offset`
  - : An [`i32`](/en-US/docs/WebAssembly/Reference/Value_types/i32) representing the offset to start writing the copied data to, in the destination memory.
- `source_offset`
  - : An `i32` representing the offset to start copying data from, in the source memory.
- `length`

>   - : An `i32` representing the number of bytes to copy.

Reviewer: All three of these would be `i64` if the memory is 64-bit (declared with `i64`). And as a fun bonus for `memory.copy`, `i32` is used for `length` when copying between a 32-bit and 64-bit memory, because the same `length` value must be used to count bytes in both 32-bit and 64-bit worlds. (It's defined in the spec as `min(addrtype(mem_1, mem_2))`, where `min(i32, i64) => i32`.)

Author: See earlier comments. I reckon doing the memory64 proposal as a separate PR would be a good idea.

Author: Update: I decided to do this after all. I've added information about it to each of these definitions — let me know what you think.

### Traps

> If the source or destination offset are out of bounds, the instruction traps.

Reviewer: This should account for the length too.

```suggestion
If any copied byte would be out of bounds in the source or destination, the instruction traps.
```

Author: Sounds good; updated

(memory.copy (i32.const 50) (i32.const 100) (i32.const 25))
```

> ### Copy specified memory

Reviewer 2: ```suggestion
### Copy within a specified memory
```

Author: done
